Long-form bootstrapping syntax allowing multiple apps, root components

alxhub opened the pull request.

alxhub commented:
This is ready for an initial review while I verify that it will actually build and run - I expect there will be some changes to get the tests passing.

alxhub commented:
Having lots of issues with the tests but debugging is yielding good results. I'm a little worried about backwards compatibility with bootstrap(), I think it will suffer, but it *is* an architectural change.

tbosch commented on modules/angular2/src/core/application_common.ts:
```diff
@@ -77,63 +77,69 @@ import {internalView} from 'angular2/src/core/compiler/view_ref';
+            (dynamicComponentLoader, injector) => {
               // TODO(rado): investigate whether to support bindings on root component.
-              return dynamicComponentLoader.loadAsRoot(appComponentType, null, injector)
+              return dynamicComponentLoader.loadAsRoot(componentType, null, injector)
```
I think this should be split up: Dont' call `DynamicComponentLoader.loadAsRoot`, but implement the logic that is in `DynamicComponentLoader.loadAsRoot` directly here:

1. call `Compiler.compileHost`, which is asynchronous
2. call `ViewManager.createRootHostView` with the result, which is synchronous

Could we have the new bootstrap adjust for this?

E.g. have an `platform.compile(componentType):Promise<ProtoViewRef>` and a `platform.application(componentType, protoViewRef, bindings):ComponentRef` and a
`platform.compileAndCreate(componentType, bindings):Promise<ComponentRef>`.

tbosch replied:
This separation is important for serverside rendering, as there we compile once, cache the result, and create the root component multiple times (on every request).

alxhub commented:
I've solved the problem I was having with `bootstrap()` backwards compatibility: `bootstrap()` accepts the user's bindings for the application, but can't determine which bindings belong in the application injector and which ones belong in the component. Some bindings, like `Router`, need to exist in the component injector because they rely on the current component, while others such as an application API service should be shared among all of an app's root components.

A solution to this problem is to introduce `Platform.simpleApplication(type, bindings)`, which returns a `Promise<RootComponentRef>` and uses a single injector but only supports one root component. `bootstrap()` then accepts any bindings and doesn't run into the same scoping problem.

mhevery commented:
@alxhub Awesome. Do you have some work in progress which you can share, so that we can start the review process?

alxhub commented:
@mhevery This PR is the full bootstrapping change that's ready for review. I'm still having some issues with the Dart tests but I can work through those separately.

alxhub committed:
feat(core): Support multiple ChangeDetectors in a single LifeCycle.

This allows a single LifeCycle to be shared among multiple root components, since each root component has its own ChangeDetector configured.

alxhub committed:
feat(core): Add a long-form syntax for Angular bootstrapping.

This change adds a syntax for bootstrapping Angular on a page that allows more fine-grained control of the hierarchy created. platform() creates a platform injector (of which there can only be one). From the platform, .application() creates an Angular application including a Zone and all specified application bindings (e.g. for the DOM, HTTP, Compiler, Renderer, etc). At the application level, .bootstrap() will bootstrap the given component into that application.

alxhub committed:
refactor(router): Switch from using APP_COMPONENT binding to a manual ROUTER_PRIMARY_COMPONENT binding.

With the coming bootstrapping changes, a single application (and thus Router) can have multiple root components. One of these needs to be identified as the "primary" component from which the Router will load its configuration. This is now done by providing a ROUTER_PRIMARY_COMPONENT binding to the primary component type.

mary-poppins commented:
User @alxhub does not have PR merging privileges.

mary-poppins commented:
User @alxhub does not have PR merging privileges.

alxhub commented:
@mhevery: Looks like I can't add the label, but it should be good to merge. Feel free to add it if you think the PR is ready to go in.

mary-poppins commented:
Merging PR #3852 on behalf of @mhevery to branch [presubmit-mhevery-pr-3852](https://github.com/angular/angular/tree/presubmit-mhevery-pr-3852).
